src/arena.c

```diff
@@ -883,6 +883,59 @@ arena_cache_bin_fill_small(tsdn_t *tsdn, arena_t *arena,
 	arena_decay_tick(tsdn, arena);
 }
 
+size_t
+arena_fill_small_fresh(tsdn_t *tsdn, arena_t *arena, szind_t binind,
+    void **ptrs, size_t nfill) {
+	if (nfill == 0) {
+		return 0;
+	}
+
+	assert(binind < SC_NBINS);
+	const bin_info_t *bin_info = &bin_infos[binind];
+	const size_t nfree = bin_info->nregs;
+	assert(nfree > 0);
+	unsigned binshard;
+	bin_t *bin = arena_bin_choose(tsdn, arena, binind, &binshard);
+	size_t nslab = 0;
+	size_t filled = 0;
+	edata_t *slab;
+
+	do {
+		slab = arena_slab_alloc(tsdn, arena, binind, binshard,
+		    bin_info);
+		size_t batch = 0;
+		if (slab != NULL) {
+			++nslab;
+			assert((size_t)edata_nfree_get(slab) == nfree);
+			batch = nfill - filled;
+			if (batch > nfree) {
+				batch = nfree;
+			}
+			assert(batch > 0);
+			arena_slab_reg_alloc_batch(slab, bin_info,
+			    (unsigned)batch, &ptrs[filled]);
+			filled += batch;
+		}
+		malloc_mutex_lock(tsdn, &bin->lock);
```

Comment:
I think it's fine for now, but in the common case in which we don't use a manual arena and don't pick sample an allocation, this does a lot of unnecessary lock acquisitions and releases. (We can instead just wait until the end, and update the stats and min slab).

Probably a premature optimization for now, but worth leaving a comment.